module/data/fields/activities-field.mjs

```diff
@@ -0,0 +1,151 @@
+ */
+export class ActivitiesField extends MappingField {
+  constructor(options) {
+    super(new ActivityField(), options);
+  }
+
+  /* -------------------------------------------- */
+
+  /** @inheritDoc */
+  initialize(value, model, options) {
+    return new ActivityCollection(model, super.initialize(value, model, options));
+  }
+}
+
+/* -------------------------------------------- */
+
+/**
+ * Field that stores activity data and swaps class based on activity type.
+ */
+export class ActivityField extends foundry.data.fields.ObjectField {
+
+  /** @override */
+  static recursive = true;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Get the document type for this activity.
+   * @param {object} value            Activity data being prepared.
+   * @returns {typeof Activity|null}  Activity document type.
+   */
+  getModel(value) {
+    return CONFIG.DND5E.activityTypes[value.type]?.documentClass ?? null;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @override */
+  _cleanType(value, options) {
+    if ( !(typeof value === "object") ) value = {};
+
+    const cls = this.getModel(value);
+    if ( cls ) return cls.cleanData(value, options);
+    return value;
+  }
+
+  /* -------------------------------------------- */
+
+  /** @override */
+  initialize(value, model, options = {}) {
+    const cls = this.getModel(value);
+    if ( cls ) return new cls(value, { parent: model, ...options });
+    return foundry.utils.deepClone(value);
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Migrate this field's candidate source data.
+   * @param {object} sourceData  Candidate source data of the root model.
+   * @param {any} fieldData      The value of this field within the source data.
+   */
+  migrateSource(sourceData, fieldData) {
+    const cls = this.getModel(fieldData);
+    if ( cls ) cls.migrateDataSafe(fieldData);
+  }
+}
+
+/* -------------------------------------------- */
+
+/**
+ * Specialized collection type for stored activities.
+ * @param {DataModel} model                   The parent DataModel to which this ActivityCollection belongs.
+ * @param {Record<string, Activity>} entries  Object containing the activities to store.
+ */
+export class ActivityCollection extends Collection {
+  constructor(model, entries) {
+    super();
+    this.#model = model;
+    for ( const [id, entry] of Object.entries(entries) ) {
+      if ( !(entry instanceof BaseActivityData) ) continue;
+      this.set(id, entry);
+    }
+  }
+
+  /* -------------------------------------------- */
+  /*  Properties                                  */
+  /* -------------------------------------------- */
+
+  /**
+   * The parent DataModel to which this ActivityCollection belongs.
+   * @type {DataModel}
+   */
+  #model;
+
+  /* -------------------------------------------- */
+
+  /**
+   * Pre-organized arrays of activities by type.
+   * @type {Map<string, Activity[]>}
```

Comment:
We could consider making this a `Map<string, Set<Activity>>` to make the `set` and `delete` implementations a little more efficient at the cost of making `getByType` a little less efficient (assuming it continues to return an array and not a `Set`).